src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,138 @@
+            head = newElement;
+        } else {
+            newElement.prev = tail;
+            tail.next = newElement;
+        }
+        tail = newElement;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        if (index == size) {
+            add(value);
+        } else if (index == 0) {
+            head.prev = new Node<>(null, value, head);
+            head = head.prev;
+            size++;
+        } else {
+            Node<T> newElement = getNodeByIndex(index);
+            Node<T> newNode = new Node<>(newElement.prev, value, newElement);
+            newElement.prev = newNode;
+            newNode.next.prev = newNode;
+            size++;
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        return getNodeByIndex(index).item;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        Node<T> newNode = getNodeByIndex(index);
+        T oldValue = newNode.item;
+        newNode.item = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        Node<T> currentNode = getNodeByIndex(index);
+        removeNode(currentNode);
+
+        return currentNode.item;
     }
 
     @Override
     public boolean remove(T object) {
+        Node<T> currentNode = getNodeByValue(object);
+
+        if (currentNode != null) {
+            removeNode(currentNode);
+            return true;
+        }
+
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private static class Node<T> {
+        private T item;
+        private Node<T> next;
+        private Node<T> prev;
+
+        Node(Node<T> prev, T element, Node<T> next) {
+            this.item = element;
+            this.next = next;
+            this.prev = prev;
+        }
+    }
+
+    private Node<T> getNodeByIndex(int index) {
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException("It's wrong index for size: " + size);
+        }
+        Node<T> currentNode = tail;
+        for (int i = size - 1; i != index; i--) {
+            currentNode = currentNode.prev;
+        }
+        return currentNode;
+    }
+
+    private Node<T> getNodeByValue(T object) {
+        Node<T> currentNode = head;
+
+        while (currentNode != null) {
+            if (currentNode.item == object || currentNode.item != null
+                    && currentNode.item.equals(object)) {
+                return currentNode;
+            }
+            currentNode = currentNode.next;
+        }
+        return null;
+    }
+
+    private void removeNode(Node<T> removeItem) {
```

Comment:
let's rename it to **unlink** to make it more clear what the purpose of the method